tsdb/tsm1/engine_schema.go

```diff
@@ -141,17 +165,33 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgBucket, tagKeyBytes
 		return true
 	})
 
+	var stats cursors.CursorStats
+
+	if canceled {
+		stats = statsFromIters(stats, iters)
+		return cursors.NewStringSliceIteratorWithStats(nil, stats), ctx.Err()
```

Comment:
I suppose this is getting a little philosophical now, but is this an error or not? `tagValuesPredicate` didn't hit any particular error conditions, it was just told to finish up early (due to deadline or explicit cancellation). It's clear here that you want to inform the caller of the (partial) work that was done. 

To me, this feels a little bit like the API should not be returning an error, but inside of the result value (the iterator), we should actually have a `Err()` method, which callers should check. That would return an error if `tagValuesPredicate` was told to terminate early.

To me, if a method returns a result and an error, and the error is `non-nil`, then the result is always undefined or useless.

Reply:
Yes I think canceling the context should result in an error. Were we to not treat it as an error, we would run into inconsistencies higher up the stack where a canceled context is treated as an error. Also as mentioned above (from a slack conversation) there is a precedent, set by the stdlib, for returning both a non-nil error as well as non-zero value result.

Having said that I will make it explicit in the godoc for TagKeys/TagValues that this is the case.